### DISCUSSION TITLE
Raise `DetectionError` instead of creating wrong model

### DISCUSSION THREAD

**Discussion by himoto** [biomass-dev/biomass#223] [Ideas]
Hi @formersbach,

Let's create a model from a text file including unregistered rule words.

Prepare `test_binding.txt` including:
```
A is binding to B --> AB
```

Text-to-model conversion
```python
from biomass import Text2Model

testmodel = Text2Model("test_binding.txt")
testmodel.convert(overwrite=True)
```
When running the code, we will get the following error message.
```
DetectionError: Unregistered words in line1: A is binding to B --> AB

Maybe: 'binds'.
```

This works as we expected.
This is because we already have a similar rule ('bind') in the reaction rules.
To make it correctly work, we need to run the following code before converting the text:

```python
testmodel.register_word({"bind": ["is binding to"]})
```

OK, let's move on to next example. Imagine that we want a model of ubiquitination.

Prepare `test_ubiquitination.txt` including:
```
A ubiquitinates B --> uB
```

Since there is no rule like 'ubiquitinate', here I want biomass to raise `DetectionError` to let users know it.
```
# This is what I expect.
DetectionError: Unregistered words in line1: A ubiquitinates B --> uB
```

However, when you run the following code:
```python
testmodel = Text2Model("test_ubiquitination.txt")
testmodel.convert(overwrite=True)
```

This does NOT raise errors but creates model.
When you look into the resulting model, you will soon find the model is wrong.
For example, in `name2idx/species.py`:
```python
from dataclasses import make_dataclass
from typing import Dict, List

NAMES: List[str] = [
    "AubiquitinatesB", # <-- this is not desired :(
    "uB",
]

...
```

It seems that this unregistered word is recognized as [`state transition rule`](https://github.com/biomass-dev/biomass/pull/169).

Do you have any ideas of how we should tackle this problem?

Ideas and PR to fix this is more than welcome!

**Comment by formersbach** [comment]
Hmm I see!
I believe this is due to line 1712 and 1713 in reaction_rules.py:
```python
                reactants = two_species[0].replace(" ", "").split("+")
                products = two_species[1].replace(" ", "").split("+")
```
For 
`
A + B --> C + D
`
these lines are necessary to remove the spaces and split at the +. 
I think this line should not be triggered if there are spaces without a + present right?
So changing to 
```python
        if len(two_species[0].split(" ") > 1 & not "+" in two_species[0]:
                reactants = two_species[0].replace(" ", "").split("+")
        if len(two_species[1].split(" ") > 1 & not "+" in two_species[1]:
                products = two_species[1].replace(" ", "").split("+")
```
should fix this. We could also prevent state_transition from triggering at all if the above mentioned conditions arise.
Do you see a better way?

**Reply by himoto** [reply]
Hi @formersbach, thanks for the idea!
> For
> A + B --> C + D
> these lines are necessary to remove the spaces and split at the +.
> I think this line should not be triggered if there are spaces without a + present right?

Yes, absolutely. This is exactly what I thought to implement. If there is no "+" symbol but a space inside reactants/products, it should be wrong.
Based on your idea & comment, I have raised an issue (#231) and created a pull request (#232) that attempted to close the issue.
I would be grateful if you could review the updates. When this is merged, I will release v0.11 that includes improvements on state_transition rule.

Thank you very much.
